### PR TITLE
add lineage rules to enforce data flow from source -> staging -> marts

### PR DESCRIPTION
**Pull request by georgim0**
adding the following rules as mentioned in #6 :
- Staging scripts referencing a staging script in a different area
- Staging scripts referencing a mart
- Marts directly referencing a source

### CODE REVIEW

File: RULES.md [hunk 1]
```diff
@@ -40,3 +39,40 @@ FROM {{ source('source_name', 'table_name') }}
 SELECT *
 FROM {{ ref('script_name') }}
 ```
+
+
+# No references outside of its staging area
+There are staging model(s) referencing a staging model that belongs in a different area
+
+If you want to cross domains, this should be done at mart level.
```
Comment (dorzey) [inline RULES.md:55]: There is mixed use of "area" and "domain". I think we should consistently use one or the other

File: RULES.md [hunk 2]
```diff
@@ -48,3 +47,40 @@ FROM {{ source('source_name', 'table_name') }}
...
+
+# No references to marts from staging
+
+There are staging model(s) referencing a mart model.
+```mermaid
```
Comment (dorzey) [inline RULES.md:60]: remove

File: RULES.md [hunk 3]
```diff
@@ -48,3 +47,40 @@ FROM {{ source('source_name', 'table_name') }}
...
+graph LR
+   source --> staging --> marts --this is bad!--> staging
+```
+![Alt text](./images/invalid_mart_reference_to_staging.png)
+
+Data should be flowing from source centric to business centric areas like so:
+```mermaid
```
Comment (dorzey) [inline RULES.md:67]: remove

File: RULES.md [hunk 4]
```diff
@@ -48,3 +47,40 @@ FROM {{ source('source_name', 'table_name') }}
...
+graph LR
+   source --> staging --> marts
+```
+![Alt text](./images/data_flow_diagram.png)
+
+# No references to source from marts
+There are mart model(s) referencing a source. 
+```mermaid
```
Comment (dorzey) [inline RULES.md:75]: remove

File: olivertwist/rules/no_references_outside_of_its_staging_area.py [hunk 5]
```diff
@@ -0,0 +1,32 @@
...
+"""
+
+from typing import List, Tuple
+
+from olivertwist.manifest import Manifest, Node
+from olivertwist.ruleengine.rule import rule
+from olivertwist.rules.utils import partition
+
+
+@rule(id="no-references-outside-of-its-staging-area", name="No references outside of its staging area")
```
Comment (dorzey) [inline olivertwist/rules/no_references_outside_of_its_staging_area.py:16]: Is it clearer if we name the rule `no-references-outside-of-own-staging-area`?

**Review by dorzey — CHANGES_REQUESTED** [review]

File: RULES.md (hunk 1 above)
Comment (georgim0) [reply RULES.md:55]: i'll go for area

File: olivertwist/rules/no_references_outside_of_its_staging_area.py (hunk 5 above)
Comment (georgim0) [reply olivertwist/rules/no_references_outside_of_its_staging_area.py:16]: how can you tell that i struggled with that? :)

**Pull request merged**